Add support for CFRunLoop integration APIs

seabaylea opened the pull request:
Work has already been done by @kovpas and merged by @parkera to add a CFRunLoop implementation for Linux into CoreFoundation by PR #133 (https://github.com/apple/swift-corelibs-foundation/pull/133).

This followed the model in CFLite of using `eventfd` rather than mach ports, and calling the following two (new) APIs:
```
dispatch_get_main_queue_eventfd_np
dispatch_main_queue_drain_np
```
instead of the existing "4CF" calls:
```
_dispatch_get_main_queue_port_4CF
_dispatch_main_queue_callback_4CF
```
Those APIs don't exist in libdispatch, so this adds them. The APIs are defined in `queue.h` which effectively makes then public - that's been done because it matched the approach used by the CoreFoundation PR. If it makes sense to make them "4CF" functions I can modify the PR here and open an equivalent against CoreFoundation.

MadCoder commented on src/queue.c:
```diff
@@ -4498,20 +4550,36 @@ _dispatch_get_main_queue_port_4CF(void)
+	_dispatch_queue_set_mainq_drain_state(false);
+}
 
-// 6618342 Contact the team that owns the Instrument DTrace probe before
```
not ok to remove a comment that says something quite explicit about it not being removed ever.

parkera replied:
This just moved up a few lines.

MadCoder commented on src/shims/pthread_main_np.h:
```diff
@@ -0,0 +1,40 @@
+static inline int
+pthread_main_np()
+{
+#if __linux__
```
this is very slow, you should instead base this patch on top of the `__thread` infrastructure that is built built.

colemancda replied:
It is indeed slow. Right now I am building a CFRunLoop based GUI in Swift, and with no input added to the runloop I am getting 0ms per loop on Mac and 17ms  on Linux. With actual input im getting up to 3000ms only on Linux, crazy.

colemancda replied:
Please check https://github.com/PureSwift/Cacao/commit/168b18d583deec82987e9ea0e84d108eac554581

MadCoder commented on dispatch/queue.h:
```diff
@@ -990,6 +990,54 @@ DISPATCH_EXPORT DISPATCH_PURE DISPATCH_WARN_RESULT DISPATCH_NOTHROW
 void *
 dispatch_get_specific(const void *key);
 
+#if __linux__
```
this should NOT be API but private interface in queue_private, and I don't understand why using the _4CF hooks isn't good enough here.

parkera replied:
Although I agree that this should not be in the public header, I do think it makes sense to make dedicated entry points for this like we also have on Windows. That means that we can change them independently from each other without having to worry about binary compatibility.

MadCoder commented:
I disagree with that patch, I think it should not add new interfaces for the sake of linux when the `_4CF` things are just fine. If you diverge too much from the Mac OS X model you will have pain keeping in sync with our changes that keep iterating on how these interfaces work.

MadCoder commented on src/queue.c:
```diff
@@ -23,6 +23,11 @@
 #include "protocol.h"
 #endif
 
+#if __linux__
```
should be in "internal.h"

MadCoder commented on src/queue.c:
```diff
@@ -77,12 +82,21 @@ static int _dispatch_pthread_sigmask(int how, sigset_t *set, sigset_t *oset);
 static void _dispatch_runloop_queue_port_dispose(dispatch_queue_t dq);
 #endif
 
+#if DISPATCH_COCOA_COMPAT || DISPATCH_LINUX_COMPAT
```
`DISPATCH_LINUX_COMPAT` is not the right name for this, it is about providing runloop backends, you should introduce a `DISPATCH_ENABLE_RUNLOOP_SUPPORT` and make sure that `DISPATCH_COCOA_COMPAT` implies it.

then fix all places that need it to use that new guard.

MadCoder commented:
For the sake of porting I think the right way is like I said, have:

1. proper guards (suggested `DISPATCH_ENABLE_RUNLOOP_SUPPORT`)
2. same function names `_4CF`, which is not a lie, it's really about giving support to CFRunloop
3. abstract internally the "port" and its use away but do not touch _dispatch_main_queue_wakeup, only what it calls. The structure should completely be the same.

seabaylea commented:
I'm happy to rename the guards and move to implementing `_4CF` functions. 

As per my initial comment, I took the approach of implementing the interface functions as  `dispatch_get_main_queue_eventfd_np` and `dispatch_main_queue_drain_np` because `CoreFoundation/RunLoop.subproj/CFRunLoop.c` had already been updated to use them.

Presumably we'll new some new `_4CF` functions for Linux as the consuming CFRunLoop code (currently) needs to know its receiving back a eventfd file descriptor rather than a "port".

seabaylea commented:
@MadCoder I've made an interaction of changes that moves towards what your looking for - they're not complete but I'd like to get some feedback before doing much more.

For the CFRunLoop entry points, I've retained retained Linux specific APIs, but moved the declaration to `queue_internal.h` and renamed them as follows:
```
int dispatch_get_main_queue_eventfd_4CF()
void _dispatch_main_queue_drain_4CF()
```
The equivalents for Darwin and Windows return a "port"/Handle, and the "drain" function for Darwin has the port as a parameter. Using shared `_4CF` functions could be achieved but would require a sizeable refactoring of the interface between CFRunLoop and libdispatch. CFRunLoop currently handles the fact that it it working with a port/Handle/eventfd for notifications whereas that would all have to be pushed down to libdispatch.
Does this seem reasonable for now? Should the definitions be moved from queue_internal.h to private.h?

I've introduced a new `DISPATCH_ENABLE_RUNLOOP_SUPPORT` guard, but had to retain the use of `DISPATCH_COCOA_COMPAT` and `DISPATCH_LINUX_COMPAT` for the Darwin/Windows and Linux specific sections. I'm not sure if this is the best approach - and feedback is appreciated.

I've also left the implementation of `pthread_main_np` as-is for the moment, as that's dependent on PR #44 which isn't yet merged.

seabaylea commented:
Hi @MadCoder . Can you take a second look at this and give some recommendations on how we can move forward?

seabaylea commented:
Hi @MadCoder . Have you been able to do another review of this yet?

MadCoder commented:
Replaced by #101
